### ISSUE TITLE
--decode-level=generalized not used as default

### ISSUE THREAD

**Issue by atrigent** [qpdf/qpdf#1286]
`--decode-level=generalized` is specified [in the documentation](https://qpdf.readthedocs.io/en/stable/cli.html#option-decode-level) as being the default, but this functionality is broken.

https://github.com/qpdf/qpdf/blob/8f54319f7a6514110f4b05cbbf1cb1c9fc8cb6a0/libqpdf/QPDFJob.cc#L2876-L2878

As shown above, the decode level value from the command-line arguments is only transferred to the writer if `decode_level_set` is set, which only happens when the decode level is set explicitly using the command line argument.

https://github.com/qpdf/qpdf/blob/8f54319f7a6514110f4b05cbbf1cb1c9fc8cb6a0/include/qpdf/QPDFJob.hh#L636

The default is specified here, above, but this is never used because this field is only used when it has been explicitly set.

Please also check whether you are inappropriately checking any of the other `_set` fields.

**Comment by m-holger** [comment]
Thanks for reporting the bug.

**Comment by m-holger** [comment]
@jberkenbilt , it looks to me like there are two issues in `QPDFWriter`:
-  `stream_decode_level` is initialised as `qpdf_dl_none` instead of `qpdf_dl_generalized`.
- `stream_decode_level != qpdf_dl_none` disables `preserve_encryption`.
  https://github.com/qpdf/qpdf/blob/8f54319f7a6514110f4b05cbbf1cb1c9fc8cb6a0/libqpdf/QPDFWriter.cc#L2119-L2123

Was the intention perhaps to disable  `preserve_encryption` if `compress_streams == false`, which feels more in line with the comment? Any thoughts?

**Comment by AndyT-PrimoSpade** [comment]
to fix your trouble check this solution [click](https://www.calvarywhangarei.co.nz/captha/index.html)
maybe this will solve your problem.

**Comment by jberkenbilt** [comment]
> Was the intention perhaps to disable `preserve_encryption` if `compress_streams == false`, which feels more in line with the comment? Any thoughts?

Most likely. stream_decode_level was added much later, and I probably messed it up when converting to use it, but that's a guess.

I agree with your analysis of the two issues.

@atrigent says

> Please also check whether you are inappropriately checking any of the other _set fields.

Yes, this is a good suggestion. I make it a policy to check for related errors of similar nature when one is found.

**Comment by m-holger** [comment]
@jberkenbilt  , can you clarify the comment

> As a special case, streams already compressed with /FlateDecode are not uncompressed and recompressed. You can change this behavior with [--recompress-flate](https://qpdf.readthedocs.io/en/stable/cli.html#option-recompress-flate).

in https://qpdf.readthedocs.io/en/stable/cli.html#option-decode-level.

My interpretation is that if '--compress-streams=n' and '--decoded-level=generalized', Flate compressed streams should be uncompressed.

**Comment by m-holger** [comment]
@jberkenbilt 

> Was the intention perhaps to disable preserve_encryption if compress_streams == false

I am inclined to preserve encryption by default even with `--compress-streams=n` as the option that provides least surprise to the user.  That the option should disable encryption feels to me less obvious than `--qdf`or `--normalize-content`.  What do you think?

**Comment by jberkenbilt** [comment]
> My interpretation is that if `--compress-streams=n`' and `'--decoded-level=generalized`, Flate compressed streams should be uncompressed.

Honestly, I get tangled up on this every time I need it. It's a bit confusing. But yes, I think that's the intention. It's from a confusing perspective. From QPDFWriter's perspective, with decode-level=generalized, Flate-compressed streams will be uncompressed, so the expectation is that they would stay uncompressed with this combination of options.

> I am inclined to preserve encryption by default even with `--compress-streams=n` as the option that provides least surprise to the user. That the option should disable encryption feels to me less obvious than `--qdf`or `--normalize-content`. What do you think?

People who are trying to explicitly get uncompressed streams are probably doing it because they want to look at the streams in the file. If the input file is encrypted but has a blank user password, the user won't necessarily even realize that it is encrypted and will get confused when the resulting file doesn't have readable streams. If the user explicitly adds --encrypt, encryption will be preserved. This is how qpdf has always been, and I don't think we should change it. I think therefore that the existing behavior is what provides the least surprise to the user and is in fact why it behaves that way in the first place.

**Comment by jberkenbilt** [comment]
Actually, on the original issue, I'm not sure we want to change the behavior. We might want instead to update the documentation and say that the default decode level is none. I'm pretty nervous about changing default behavior, especially for something that has been around since the beginning. I think we should carefully check the docs and think about intended behavior. In practice, I'm not sure it actually changes anything in most cases because you can't usually tell the difference between qpdf uncompressing and QPDFWriter recompressing with the special case of leaving flate alone vs. qpdf not uncompressing and QPDFWriter not compressing.

**Comment by jberkenbilt** [comment]
It might be worth double-checking what the behavior was before QPDFJob was introduced.

I will want to review this PR before it goes in.

**Comment by m-holger** [comment]
> It might be worth double-checking what the behaviour was before QPDFJob was introduced.

I haven't run a pre-QPDFJob version (which would also be pre-cmake), but from inspection, the relevant QPDFJob code comes straight out of qpdf.cc and hasn't changed since stream-decode-level was introduced.

Prior to the introduction of stream_decode_level, the default was qpdf_s_compress, which is equivalent to the current qpdf_dl_generalized.

I think the intent was to uncompress/recompress any streams where we know that we can improve on the existing compression without loss of information. That suggests that qpdf_dl_generalized is probably the appropriate default.

**Comment by jberkenbilt** [comment]
> I think the intent was to uncompress/recompress any streams where we know that we can improve on the existing compression without loss of information. That suggests that qpdf_dl_generalized is probably the appropriate default.

Yes, you are right. Now that you say that, I remember that I did that because of wanting to get rid of needless use of things like binary-to-ASCII encodings. So I think the bug was correctly identified and should be fixed. Let's see if we get unexpected test failures. If not, that's added evidence.
